File: poo/python/HerancaAbstrata/Banco.py
#!/usr/bin/env python
#-*- coding: utf-8 -*-

'''
@author Nathan Cruz <nathan.cruz@unifesp.br>
@created 13/04/14
'''

from __future__ import unicode_literals
from __future__ import print_function

from ContaCorrente import ContaCorrente
from ContaPoupanca import ContaPoupanca


class Banco(object):

    # Inicializador
    def __init__(self):
        self.__dinheiro = 0.0
        self.__contas = []


    # Registra novas contas
    def Registra(self, conta):
        self.__dinheiro += conta.getSaldo()


    # Imprime o total do banco
    def Imprime(self):
        print('{0} {1}'.format('Saldo do banco:', self.__dinheiro))


if __name__ == '__main__':

    banco = Banco()
    contas = []

    for i in range(10):
        if i % 2:
            contas.append(ContaCorrente(i))
        else:
            contas.append(ContaPoupanca(i))

    banco.contas = contas

    for i in range(10):
        banco.contas[i].credita(12)
        banco.contas[i].debita(3)
        banco.Registra(banco.contas[i])

    banco.Imprime()

File: poo/java/Exceptions/Exercicio6/Banco.java
/*
 *  Programa escrito com propósito didático apenas.
 *  @author Nathan Cruz <nathan.cruz@unifesp.br>
 *  @created 04/04/14
 *  R.A.: 86492
 */

import java.util.*;

public class Banco{

    /* Atributos */
    private double dinheiro;
    private HashSet<Conta> contas = new HashSet<Conta>();

    /* Construtor */
    public Banco(){
        this.dinheiro = 0.0;
    }

    /* Registra nova conta */
    public void Registra(Conta conta) throws ExcecaoElementoJaExistente{
        if(this.contas.add(conta)){
            this.dinheiro += conta.getSaldo();
        }
        else{
            throw new ExcecaoElementoJaExistente("Conta ja cadastrada");
        }
    }

    /* Imprime o total do banco */
    public void Imprime(){

        /* Atualiza o valor de this.dinheiro antes de imprimir */
        for(Conta conta: this.contas){
            this.dinheiro += conta.getSaldo();
        }
        System.out.println("Saldo do banco: " + this.dinheiro);
    }

    public static void main(String args[]){
        Banco banco = new Banco();
        ContaCorrente[] cc = new ContaCorrente[5];
        ContaPoupanca[] cp = new ContaPoupanca[5];
        int i, j;

        for(i=0;i<5;i++){
            cc[i] = new ContaCorrente(i);
            cp[i] = new ContaPoupanca(i);
        }

        for(i=0;i<5;i++){
            try{
                banco.Registra(cc[2]);
                banco.Registra(cp[i]);
            }
            catch(ExcecaoElementoJaExistente e){
                System.out.println(e.getMessage());
            }
        }

        for(Conta conta: banco.contas){
            try{
                conta.credita(12);
                conta.debita(13);
            }
            catch(ValorInvalidoException e){
                System.out.println(e.getMessage());
            }
        }

        banco.Imprime();

    }
}
